### PR TITLE
[MTL] non-low power mode support for vp9 8bit encode

### PR DESCRIPTION
**Pull request by FocusLuo**
there will no lp mode for MTL encode
add non-lp mode support for vp9 8bit encode for ffmpeg and gstreamer

### CODE REVIEW

File: lib/ffmpeg/encoderbase.py [hunk 1]
```diff
@@ -141,6 +141,16 @@ def before(self):
...
+  def get_lowpower_opt(self):
+    if get_media()._get_gpu_gen() < 13:
+         #only cover "low_power=0" path
+         return 0
+    else:
+         #For MTL platforms (gen >== 13)
+         #will add random() to set "low_power=0" and no "low_power" option
+         #to cover more test path in future
+         return 0
+
```
Comment (uartie) [inline lib/ffmpeg/encoderbase.py:153]: I don't like this solution.  It's confusing.  Why can't we just let MTL+ use the normal lowpower=0 in the tests to start.

File: lib/ffmpeg/qsv/encoder.py [hunk 2]
```diff
@@ -121,7 +121,7 @@ def check_output(self):
       assert m is not None, "Possible incorrect RC mode used"
 
     # lowpower
-    if self.codec not in ["jpeg", "mpeg2"]:
+    if self.codec not in ["jpeg", "mpeg2", "vp9", "vp9-10", "av1", "av1-10"]:
```
Comment (uartie) [inline lib/ffmpeg/qsv/encoder.py:124]: Why?  Now we will be skipping < MTL check.

File: lib/ffmpeg/qsv/encoder.py [hunk 3]
```diff
@@ -380,6 +380,16 @@ def before(self):
   def get_file_ext(self):
     return "ivf"
 
+@slash.requires(*platform.have_caps("encode", "vp9_8"))
+class VP9_8EncoderTest(VP9_8EncoderBaseTest):
+  def before(self):
+    super().before()
+    _lowpower_ = self.get_lowpower_opt()
```
Comment (uartie) [inline lib/ffmpeg/qsv/encoder.py:387]: Let's avoid this solution for now.  Just use lowpower=0 to start.

File: lib/ffmpeg/vaapi/encoder.py [hunk 4]
```diff
@@ -66,9 +66,15 @@ def check_output(self):
     assert m is not None, "Possible incorrect profile used"
 
     # entrypoint
+    if "jpeg" == self.codec:
+      _VAEntrypoint_Encode_Mode_ = "VAEntrypointEncPicture"
+    elif self.codec in ["vp9"]:
+      _VAEntrypoint_Encode_Mode_ = "VAEntrypointEncSliceLP"
+    else :
+      _VAEntrypoint_Encode_Mode_ = "VAEntrypointEncSlice"
```
Comment (uartie) [inline lib/ffmpeg/vaapi/encoder.py:74]: Too confusing!

File: lib/ffmpeg/vaapi/encoder.py [hunk 5]
```diff
@@ -92,7 +98,7 @@ def check_output(self):
     ipbmode = 0 if vars(self).get("gop", 0) <= 1 else 1 if vars(self).get("bframes", 0) < 1 else 2
     ipbmsgs = [
       "Using intra frames only",
-      "Using intra and P-frames|[L|l]ow delay|forward-prediction"
+      "|Using intra and P-frames|[L|l]ow delay|forward-prediction"
```
Comment (uartie) [inline lib/ffmpeg/vaapi/encoder.py:101]: Why?  This is a new entry in the list... why put an `|` operator at the start?

**Review by uartie — CHANGES_REQUESTED** [review]
I think we can do better.  This just feels like hackery.

File: lib/ffmpeg/vaapi/encoder.py (hunk 4 above)
Comment (uartie) [reply lib/ffmpeg/vaapi/encoder.py:74]: This also breaks legacy

File: lib/ffmpeg/encoderbase.py (hunk 1 above)
Comment (FocusLuo) [reply lib/ffmpeg/encoderbase.py]: ok, removed

File: lib/ffmpeg/qsv/encoder.py (hunk 2 above)
Comment (FocusLuo) [reply lib/ffmpeg/qsv/encoder.py]: thanks for finding this, it does not need

File: lib/ffmpeg/vaapi/encoder.py (hunk 4 above)
Comment (FocusLuo) [reply lib/ffmpeg/vaapi/encoder.py]: Removed. Right, this is an issue, thanks finding this.

File: lib/ffmpeg/vaapi/encoder.py (hunk 5 above)
Comment (FocusLuo) [reply lib/ffmpeg/vaapi/encoder.py]: Removed. This is not needed

**Comment by FocusLuo** [comment]
MTL ffmpeg-qsv vp9 encode log
[2023-09-26 17:20:48] [vp9_qsv @ 0x5562b48cf0c0] Using input frames context (format qsv) with vp9_qsv encoder.
[2023-09-26 17:20:48] [vp9_qsv @ 0x5562b48cf0c0] Encoder: input is video memory surface
[2023-09-26 17:20:48] [vp9_qsv @ 0x5562b48cf0c0] Use Intel(R) oneVPL to create MFX session with the specified MFX loader
[2023-09-26 17:20:48] [vp9_qsv @ 0x5562b48cf0c0] Using the constant bitrate (CBR) ratecontrol method
[2023-09-26 17:20:48] [vp9_qsv @ 0x5562b48cf0c0] profile: vp9 0
[2023-09-26 17:20:48] [vp9_qsv @ 0x5562b48cf0c0] GopPicSize: 30; GopRefDist: 1; GopOptFlag:; IdrInterval: 0
[2023-09-26 17:20:48] [vp9_qsv @ 0x5562b48cf0c0] TargetUsage: 4; RateControlMethod: CBR
[2023-09-26 17:20:48] [vp9_qsv @ 0x5562b48cf0c0] BufferSizeInKB: 1000; InitialDelayInKB: 500; TargetKbps: 4000; MaxKbps: 4000; BRCParamMultiplier: 1
[2023-09-26 17:20:48] [vp9_qsv @ 0x5562b48cf0c0] NumRefFrame: 2
[2023-09-26 17:20:48] [vp9_qsv @ 0x5562b48cf0c0] FrameRateExtD: 1; FrameRateExtN: 30
[2023-09-26 17:20:48] [vp9_qsv @ 0x5562b48cf0c0] IntRefType: 0; IntRefCycleSize: 0; IntRefQPDelta: 0
[2023-09-26 17:20:48] [vp9_qsv @ 0x5562b48cf0c0] MaxFrameSize: 0
[2023-09-26 17:20:48] [vp9_qsv @ 0x5562b48cf0c0] BitrateLimit: unknown; MBBRC: OFF; ExtBRC: unknown
[2023-09-26 17:20:48] [vp9_qsv @ 0x5562b48cf0c0] **### VDENC: ON**
[2023-09-26 17:20:48] [vp9_qsv @ 0x5562b48cf0c0] MinQPI: 0; MaxQPI: 0; MinQPP: 0; MaxQPP: 0; MinQPB: 0; MaxQPB: 0
[2023-09-26 17:20:48] [vp9_qsv @ 0x5562b48cf0c0] WriteIVFHeaders: OFF
[2023-09-26 17:20:48] Automatically inserted bitstream filter 'vp9_superframe'; args=''
[2023-09-26 17:20:48] Output #0, ivf, to '/tmp_sda4/work_focus/applications.media.middleware.validation.vaapi-fits-smoke/results/smoke/07c083d6-5c91-11ee-b8c0-b53dd3ceca9b_0/_0.test.ffmpeg-qsv.encode.vp9/cbr/07c083d6-5c91-11ee-b8c0-b53dd3ceca9b_0_1_1.ivf':
[2023-09-26 17:20:48]   Metadata:
[2023-09-26 17:20:48]     encoder         : Lavf60.10.100
[2023-09-26 17:20:48]   Stream #0:0: Video: vp9, 1 reference frame (VP90 / 0x30395056), qsv(tv, progressive), 704x576 (0x0), q=2-31, 4000 kb/s, 30 fps, 30 tbn
[2023-09-26 17:20:48]     Metadata:
[2023-09-26 17:20:48]       encoder         : Lavc60.23.100 vp9_qsv
[2023-09-26 17:20:48] frame=    0 fps=0.0 q=0.0 size=       0kB time=N/A bitrate=N/A speed=N/A
[2023-09-26 17:20:48] No more output streams to write to, finishing.
[2023-09-26 17:20:48] [vist#0:0/rawvideo @ 0x5562b48210c0] Decoder thread received EOF packet
[2023-09-26 17:20:48] [vist#0:0/rawvideo @ 0x5562b48210c0] Decoder returned EOF, finishing
[2023-09-26 17:20:48] [vist#0:0/rawvideo @ 0x5562b48210c0] Terminating decoder thread
[2023-09-26 17:20:48] [out#0/ivf @ 0x5562b4854340] All streams finished
[2023-09-26 17:20:48] [out#0/ivf @ 0x5562b4854340] Terminating muxer thread
[2023-09-26 17:20:48] [AVIOContext @ 0x5562b48d2d80] Statistics: 2481856 bytes written, 2 seeks, 11 writeouts
[2023-09-26 17:20:48] [out#0/ivf @ 0x5562b4854340] Output file #0 (/tmp_sda4/work_focus/applications.media.middleware.validation.vaapi-fits-smoke/results/smoke/07c083d6-5c91-11ee-b8c0-b53dd3ceca9b_0/_0.test.ffmpeg-qsv.encode.vp9/cbr/07c083d6-5c91-11ee-b8c0-b53dd3ceca9b_0_1_1.ivf):
[2023-09-26 17:20:48] [out#0/ivf @ 0x5562b4854340]   Output stream #0:0 (video): 150 frames encoded; 150 packets muxed (2480020 bytes);
[2023-09-26 17:20:48] [out#0/ivf @ 0x5562b4854340]   Total: 150 packets (2480020 bytes) muxed
[2023-09-26 17:20:48] [out#0/ivf @ 0x5562b4854340] video:2422kB audio:0kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: 0.073870%
[2023-09-26 17:20:48] frame=  150 fps=0.0 q=-0.0 Lsize=    2424kB time=00:00:04.96 bitrate=3997.6kbits/s speed=10.4x
[2023-09-26 17:20:48] [in#0/rawvideo @ 0x5562b4820040] Terminating demuxer thread
[2023-09-26 17:20:48] [in#0/rawvideo @ 0x5562b4820040] Input file #0 (/tmp_sda4/work_focus/assets/otc-media/sanity/yuv/SOCCER_704x576_30_orig_02_704x576_I420.yuv):
[2023-09-26 17:20:48] [in#0/rawvideo @ 0x5562b4820040]   Input stream #0:0 (video): 152 packets read (92454912 bytes); 150 frames decoded; 0 decode errors;
[2023-09-26 17:20:48] [in#0/rawvideo @ 0x5562b4820040]   Total: 152 packets (92454912 bytes) demuxed
[2023-09-26 17:20:48] [AVIOContext @ 0x5562b4862940] Statistics: 92454912 bytes read, 0 seeks
[2023-09-26 17:20:48] DETAIL: time(ffmpeg:encode:1) = 0.5360s

**Comment by FocusLuo** [comment]
user@P12SL03SLIM50:/tmp_sda4/work_focus/applications.media.middleware.validation.vaapi-fits-smoke$ ./smoke run test/gst-va/encode/10bit/vp9.py --pl MTL --context=sanity -vv; ./smoke run test/gst-msdk/encode/10bit/vp9.py --pl MTL --context=sanity -vv; ./smoke run test/ffmpeg-vaapi/encode/10bit/vp9.py --pl MTL --context=sanity -vv; ./smoke run test/ffmpeg-qsv/encode/10bit/vp9.py --pl MTL --context=sanity -vv;
all passed

**Comment by FocusLuo** [comment]
user@P12SL03SLIM50:/tmp_sda4/work_focus/vaapi-fits$ vainfo
Trying display: wayland
Trying display: x11
error: can't connect to X server!
Trying display: drm
libva info: VA-API version 1.20.0
libva info: User environment variable requested driver 'iHD'
libva info: Trying to open /tmp_sda4/work_focus/media/build/lib/dri/iHD_drv_video.so
libva info: Found init function __vaDriverInit_1_20
libva info: va_openDriver() returns 0
vainfo: VA-API version: 1.20 (libva 2.20.0)
vainfo: Driver version: Intel iHD driver for Intel(R) Gen Graphics - 23.3.4 (94542dbcb)
vainfo: Supported profile and entrypoints
      VAProfileNone                   : VAEntrypointVideoProc
      VAProfileMPEG2Simple            : VAEntrypointVLD
      VAProfileMPEG2Main              : VAEntrypointVLD
      VAProfileH264Main               : VAEntrypointVLD
      VAProfileH264Main               : VAEntrypointEncSlice
      VAProfileH264High               : VAEntrypointVLD
      VAProfileH264High               : VAEntrypointEncSlice
      VAProfileJPEGBaseline           : VAEntrypointVLD
      VAProfileJPEGBaseline           : VAEntrypointEncPicture
      VAProfileH264ConstrainedBaseline: VAEntrypointVLD
      VAProfileH264ConstrainedBaseline: VAEntrypointEncSlice
      VAProfileVP8Version0_3          : VAEntrypointVLD
      VAProfileHEVCMain               : VAEntrypointVLD
      VAProfileHEVCMain               : VAEntrypointEncSlice
      VAProfileHEVCMain10             : VAEntrypointVLD
      VAProfileHEVCMain10             : VAEntrypointEncSlice
      VAProfileVP9Profile0            : VAEntrypointVLD
      VAProfileVP9Profile0            : VAEntrypointEncSlice
      VAProfileVP9Profile1            : VAEntrypointVLD
      VAProfileVP9Profile1            : VAEntrypointEncSlice
      VAProfileVP9Profile2            : VAEntrypointVLD
      VAProfileVP9Profile2            : VAEntrypointEncSlice
      VAProfileVP9Profile3            : VAEntrypointVLD
      VAProfileVP9Profile3            : VAEntrypointEncSlice
      VAProfileHEVCMain12             : VAEntrypointVLD
      VAProfileHEVCMain422_10         : VAEntrypointVLD
      VAProfileHEVCMain422_12         : VAEntrypointVLD
      VAProfileHEVCMain444            : VAEntrypointVLD
      VAProfileHEVCMain444            : VAEntrypointEncSlice
      VAProfileHEVCMain444_10         : VAEntrypointVLD
      VAProfileHEVCMain444_10         : VAEntrypointEncSlice
      VAProfileHEVCMain444_12         : VAEntrypointVLD
      VAProfileHEVCSccMain            : VAEntrypointVLD
      VAProfileHEVCSccMain            : VAEntrypointEncSlice
      VAProfileHEVCSccMain10          : VAEntrypointVLD
      VAProfileHEVCSccMain10          : VAEntrypointEncSlice
      VAProfileHEVCSccMain444         : VAEntrypointVLD
      VAProfileHEVCSccMain444         : VAEntrypointEncSlice
      VAProfileAV1Profile0            : VAEntrypointVLD
      VAProfileAV1Profile0            : VAEntrypointEncSlice
      VAProfileHEVCSccMain444_10      : VAEntrypointVLD
      VAProfileHEVCSccMain444_10      : VAEntrypointEncSlice

File: test/ffmpeg-qsv/encode/vp9.py [hunk 6]
```diff
@@ -68,3 +67,61 @@ def test(self, case, gop, bitrate, fps, slices, refmode, quality, looplvl, loops
...
+    vars(self).update(
+      case      = case,
+      gop       = 30 if ipmode != 0 else 1,
+      qp        = qp,
+      slices    = slices,
+      rcmode    = "cqp",
+      quality   = quality,
+    )
+
+  @parametrize_with_unused(*gen_vp9_cqp_lp_parameters(spec), ['refmode', 'loopshp', 'looplvl'])
```
Comment (uartie) [inline test/ffmpeg-qsv/encode/vp9.py:82]: Oops, I missed this one.  We need a new case generator for "cqp" variants.  Please submit a new PR to fix these.

**Pull request merged**